acra/configuring-maintaining/optimisations/acrastructs_vs_acrablocks.md

```diff
@@ -0,0 +1,27 @@
+---
+title: AcraStructs vs AcraBlocks
+weight: 3
+---
+
+# AcraStructs vs AcraBlocks
+
+There are two different "crypto containers" AcraServer can use to store encrypted value in.
+Then, these containers are stored in database and decrypted back by AcraServer when needed.
+
+History
+* `AcraStruct` is our "old" one, though we're not going to remove it any time soon.
+* `AcraBlock` on the other hand is new and proposes some extra features.
+
+Encryption with a key
+* `AcraStruct` uses key provided from somewhere outside (by Acra) to encrypt the value.
+* `AcraBlock` uses random key to encrypt the value, then encrypts this key with another key, this time provided from outside.
+  That first key is called DEK (data encryption key) and the second one is called KEK (key encryption key).
+
+Key rotation (when you need re-encrypt data with a new key)
+* With `AcraStruct`, the value has to be decrypted completely, then encrypted with a new key.
+* With `AcraBlock`, only DEK should be re-encrypted, leaving the encrypted value the same.
+  And this may have a big impact on performance when rotating keys for a large amount of encrypted data.
+
+Storage overhead
+* `AcraStruct` uses a bit less extra bytes compared to `AcraBlock`.
```

Comment:
hm.... i'm not sure that AcraStructs uses less than AcraBlock.
```
package main

import (
	"fmt"
	"github.com/cossacklabs/acra/acrablock"
	"github.com/cossacklabs/acra/acrastruct"
	"github.com/cossacklabs/themis/gothemis/keys"
)
import _ "github.com/lib/pq"

func main() {
	testData := []byte(`a`)
	ecKeys, err := keys.New(keys.TypeEC)
	if err != nil {
		panic(err)
	}
	acraStruct, err := acrastruct.CreateAcrastruct(testData, ecKeys.Public, nil)
	if err != nil {
		panic(err)
	}
	acraBlock, err := acrablock.CreateAcraBlock(testData, []byte(`some key`), nil)
	if err != nil {
		panic(err)
	}
	fmt.Println(len(acraStruct), len(acraBlock))
}
// >> 190 139
```

AcraStruct: 190
AcraBlock 139

Reply:
Changed, is new version better?
(talking about storage overhead comparison, other issues in this file will be fixed in next commit)